Refactor telemetry code

datenreisender opened the pull request:
A bunch of enhancements (hopefully), done while I was teaching myself how the code works. The biggest change is the last commit 0b288e20495f4df2962386006598d9df88747937, which introduces classes in the telemetry code because to me this seems to suit the structure with two variants (being either in a renderer or in the main process) and some shared logic better.

There are some additional changes I would like to do but they are not refactorings but instead changing the API. So I did those in the separate PR #837.

kylebonnici commented on src/telemetry/TelemetrySenderInRenderer.ts:
```diff
@@ -0,0 +1,82 @@
+        this.client.loadAppInsights();
+
+        // Add app name and version to every event
+        this.client.addTelemetryInitializer(envelope => {
+            if (envelope.data?.removeAllMetadata) {
+                envelope.data = {};
+                envelope.baseData = { name: envelope.baseData?.name };
+                envelope.ext = {};
+                envelope.tags = [];
+            } else {
+                const trace = {
+                    ...(envelope.ext?.trace ?? {}),
+                    name: applicationName,
+                };
+                envelope.ext = { ...envelope.ext, trace };
+                envelope.baseData = {
+                    applicationName,
+                    applicationVersion,
+                    isDevelopment,
+                    source,
+                    ...envelope.baseData,
+                };
+            }
+        });
+
+        return this.client;
+    }
```
```suggestion
        this.client.addTelemetryInitializer(envelope => {
            if (envelope.data?.removeAllMetadata) {
                envelope.data = {};
                envelope.baseData = { name: envelope.baseData?.name };
                envelope.ext = {};
                envelope.tags = [];
            } else {
                const trace = {
                    ...(envelope.ext?.trace ?? {}),
                    name: applicationName,
                };
                envelope.ext = { ...envelope.ext, trace };
                envelope.data= {
                    ...envelope.data,,
                    applicationName,
                    applicationVersion,
                    isDevelopment,
                    source
                };
            }
        });

        return this.client;
    }
```

kylebonnici replied:
This was a fix I had introduced

kylebonnici commented on src/telemetry/TelemetrySenderInRenderer.ts:
```diff
@@ -0,0 +1,82 @@
+
+        const accountId = getTelemetryClientId();
+
+        this.client = new ApplicationInsights({
+            config: {
+                instrumentationKey: this.INSTRUMENTATION_KEY,
+                accountId, // to hide with removeAllMetadata
+            },
+        });
```
```suggestion
        this.client = new ApplicationInsights({
            config: {
                instrumentationKey: this.INSTRUMENTATION_KEY,
                accountId, // to hide with removeAllMetadata
                 isStorageUseDisabled: true, // fix issue with duplicate events being sent https://github.com/microsoft/ApplicationInsights-JS/issues/796
                 namePrefix: applicationName, // fix issue with duplicate events being sent https://github.com/microsoft/ApplicationInsights-JS/issues/796
            },
        });
```

kylebonnici replied:
Avoid duplicate events

kylebonnici commented on src/telemetry/TelemetrySenderInMain.ts:
```diff
@@ -0,0 +1,76 @@
+        this.client.config.enableAutoDependencyCorrelation = false;
+
+        // Add app name and version to every event
+        this.client.addTelemetryProcessor(envelope => {
+            if (envelope.data.baseData?.properties.removeAllMetadata) {
+                envelope.tags = [];
+                envelope.data.baseData = {
+                    name: envelope.data?.baseData.name,
+                };
+            } else {
+                envelope.tags['ai.cloud.roleInstance'] = undefined; // remove PC name
+                envelope.data.baseData = {
+                    applicationName,
+                    applicationVersion,
+                    isDevelopment,
+                    ...envelope.data.baseData,
+                };
+            }
+
+            return true;
+        });
```
```suggestion
        this.client.addTelemetryProcessor(envelope => {
            if (envelope.data.baseData?.properties.removeAllMetadata) {
            envelope.tags = [];
            envelope.data.baseData = { name: envelope.data?.baseData.name };
        } else {
            envelope.tags['ai.cloud.roleInstance'] = undefined; // remove PC name
            if (envelope.data.baseData) {
                envelope.data.baseData.properties = {
                    applicationName,
                    applicationVersion,
                    isDevelopment,
                    ...envelope.data.baseData.properties,
                };
            }
        }

        return true;
        });
```

datenreisender commented:
@kylebonnici You were absolutely right that I forgot the changes which we did in the telemetry code after I started this refactoring. Thanks for noticing. I now incorporated those:
- Avoid duplicate events https://github.com/NordicSemiconductor/pc-nrfconnect-shared/blob/35e2d56ccc3632ba13133ebbf49f4f3956b6c33f/src/telemetry/TelemetrySenderInRenderer.ts#L31-L32
- Corrected props https://github.com/NordicSemiconductor/pc-nrfconnect-shared/blob/35e2d56ccc3632ba13133ebbf49f4f3956b6c33f/src/telemetry/TelemetrySenderInRenderer.ts#L51-L57 and https://github.com/NordicSemiconductor/pc-nrfconnect-shared/blob/35e2d56ccc3632ba13133ebbf49f4f3956b6c33f/src/telemetry/TelemetrySenderInMain.ts#L44-L49

I checked all changes done to the telemetry code again and also noticed that I forgot the fix from #846 and also copied that over: https://github.com/NordicSemiconductor/pc-nrfconnect-shared/blob/35e2d56ccc3632ba13133ebbf49f4f3956b6c33f/src/telemetry/telemetry.ts#L28-L33

I think now everything should be fine.
